### PR TITLE
Added support for user creation and deletion

### PR DESCRIPTION
**Pull request by saurav-duplo**
### **PR Type**
enhancement, tests

___

### **Description**
- Added support for creating and deleting users, reflected in the CHANGELOG.
- Introduced new test data files for creating an admin user, a non-admin user with extensive properties, and deleting a user.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update CHANGELOG for User Management Features</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
CHANGELOG.md

<li>Added entry for support of user creation and deletion in the <br>changelog.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/75/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>create-admin-user.yaml</strong><dd><code>Add Test Data for Creating Admin User</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/tests/data/create-admin-user.yaml

<li>Added a new test data file for creating an admin user with specified <br>roles and state.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/75/files#diff-8e5f7d12aeaa8e2512831e95ca27128223d694e2c16623f126283a7d3a20e199">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>create-non-admin-user.yaml</strong><dd><code>Add Test Data for Creating Non-Admin User</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/tests/data/create-non-admin-user.yaml

<li>Added a new test data file for creating a non-admin user with multiple <br>roles and properties.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/75/files#diff-30da7e9e60a53f22d985f7f6ed5842bd15951f37123cc1a8911d6986fa694c21">+16/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>delete-user.yaml</strong><dd><code>Add Test Data for Deleting User</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
src/tests/data/delete-user.yaml

<li>Added a new test data file for deleting a user with specified username <br>and state.<br>

</details>
    

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/75/files#diff-2e3e3181de71d80997607f62f502704df5beb613de5c7759ce04e93bb822d6a1">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by zafarabbas** [comment]
Task linked: [DUPLO-17365 create users in duplocloud console](https://app.clickup.com/t/8655600/DUPLO-17365)

File: src/tests/data/create-admin-user.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+Username: test@gmail.com
+Roles:
+- Administrator
+State: added
```
Comment (kferrone) [inline src/tests/data/create-admin-user.yaml:5]: Is `state` really an input? Can this be omitted or does the api actually force this to be present?

File: src/tests/data/create-non-admin-user.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+---
+Username: test2@gmail.com
+Roles:
+- User
+- SignupUser
+VpnPassword: "********"
```
Comment (kferrone) [inline src/tests/data/create-non-admin-user.yaml:6]: Same question as above, isn't `vpnpassword` readonly? Is this really part of the input?

**Review by kferrone — COMMENTED** [review]
Where are the code changes?

**Comment by saurav-duplo** [comment]
> Where are the code changes?

Pushed now please verify

File: src/tests/data/create-admin-user.yaml (hunk 1 above)
Comment (saurav-duplo) [reply src/tests/data/create-admin-user.yaml:5]: This was present in request body of the api call made from UI

File: src/tests/data/create-non-admin-user.yaml (hunk 2 above)
Comment (saurav-duplo) [reply src/tests/data/create-non-admin-user.yaml:6]: This was present in request body of the api call made from UI

File: src/duplo_resource/user.py [hunk 3]
```diff
@@ -41,3 +41,10 @@ def add_user_to_tenant(self,
       raise DuploError(f"Failed to add user '{name}' to tenant '{tenant}'", res["status_code"])
     else:
       return f"User '{name}' added to tenant '{tenant}'"
+    
+  @Command()
+  def create(self,
```
Comment (kferrone) [inline src/duplo_resource/user.py:46]: I see what's up with the `state` key in the body now. 
You can instead always make sure `state` is `add` in `create`. 
Like: `body['state'] = 'add'` 
Then make a delete method which only takes the name and make the body with `state` as `delete`. This way we ensure the common CRUD methods are present.

**Review by kferrone — CHANGES_REQUESTED** [review]

**Review by kferrone — APPROVED** [review]

**Pull request merged**